packages/contractkit/src/providers/celo-private-keys-subprovider.ts

```diff
@@ -0,0 +1,192 @@
+import { Callback, ErrorCallback, PrivateKeyWalletSubprovider } from '@0x/subproviders'
+import debugFactory from 'debug'
+import { JSONRPCRequestPayload } from 'ethereum-types'
+import Web3 from 'web3'
+import { signTransaction } from '../utils/signing-utils'
+import { CeloPartialTxParams } from '../utils/tx-signing'
+
+const debug = debugFactory('kit:providers:celo-private-keys-subprovider')
+
+function getPrivateKeyWithout0xPrefix(privateKey: string) {
+  return privateKey.toLowerCase().startsWith('0x') ? privateKey.substring(2) : privateKey
+}
+
+export function generateAccountAddressFromPrivateKey(privateKey: string): string {
+  if (!privateKey.toLowerCase().startsWith('0x')) {
+    privateKey = '0x' + privateKey
+  }
+  return new Web3().eth.accounts.privateKeyToAccount(privateKey).address
+}
+
+/**
+ * This class supports storing multiple private keys for signing.
+ * The base class PrivateKeyWalletSubprovider only supports one key.
+ */
+export class CeloPrivateKeysWalletProvider extends PrivateKeyWalletSubprovider {
+  // Account addresses are hex-encoded, lower case alphabets
+  private readonly accountAddressToPrivateKey = new Map<string, string>()
+
+  private chainId: number | null = null
+  private gasFeeRecipient: string | null = null
+
+  constructor(privateKey: string) {
+    // This won't accept a privateKey with 0x prefix and will call that an invalid key.
+    super(getPrivateKeyWithout0xPrefix(privateKey))
+    this.addAccount(privateKey)
+  }
+
+  public addAccount(privateKey: string) {
+    // Prefix 0x here or else the signed transaction produces dramatically different signer!!!
+    privateKey = '0x' + getPrivateKeyWithout0xPrefix(privateKey)
```

Comment:
celo/utils has or should have an `ensure0x` (if it's not in utils it's celotool and should be ported)

Reply:
It is in celotool and it is broken `export const ensure0x = (hexstr: string) => (hexstr.startsWith('0x') ? hexstr : '0x' + hexstr)` since it will miss a string which starts with `0X`. The consquence of that would be a disaster. Let's fix that in a separate PR, so that, 5 unrelated files are not impacted.

```
 $ rg ensure0x
celotool/src/lib/generate_utils.ts
12:import { ensure0x, strip0x } from '@celo/celotool/src/lib/utils'
77:  return new Web3.modules.Eth().accounts.privateKeyToAccount(ensure0x(privateKey)).address

celotool/src/lib/utils.ts
146:export const ensure0x = (hexstr: string) => (hexstr.startsWith('0x') ? hexstr : '0x' + hexstr)

celotool/geth_tests/src/lib/utils.ts
11:import { ensure0x } from '@celo/celotool/src/lib/utils'
333:    validatorPrivateKeys.map(ensure0x).join(','),

celotool/src/cmds/geth/get_gold_balance.ts
5:  ensure0x,
41:  account = ensure0x(account)

celotool/src/cmds/geth/run.ts
5:import { addCeloGethMiddleware, ensure0x, validateAccountAddress } from 'src/lib/utils'
80:      (minerAddress: string) => (minerAddress === null ? null : ensure0x(minerAddress))

celotool/src/cmds/deploy/initial/contracts.ts
6:import { ensure0x, execCmd } from '@celo/celotool/src/lib/utils'
24:    return ensure0x(OG_ACCOUNTS[0].address)
34:    return OG_ACCOUNTS.map((account) => account.privateKey).map(ensure0x)
40:    ).map(ensure0x)
```

Reply:
sure, let's do that! do you want to create and issue or do that after this pr?

Reply:
I will send a follow-on PR to fix `ensure0x`. Let's not worry about de-duplicating this function for now.